# HW2

Jaee Oh / Tejas Phanse / Mandadi Ayush Reddy

## Part A: DTM

In [1]:
import nltk
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer
import re
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer
from bs4 import BeautifulSoup
from sklearn.feature_extraction.text import TfidfVectorizer
import nltk
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import cross_validate, StratifiedKFold
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import make_scorer, precision_score, recall_score, f1_score

nltk.download('stopwords')
nltk.download('wordnet') 
print("Libraries loaded successfully.")

Libraries loaded successfully.


[nltk_data] Downloading package stopwords to /home/jaee/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package wordnet to /home/jaee/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


- Text Prerpcessing

In [2]:
stop_words = set(stopwords.words('english'))
lemmatizer = WordNetLemmatizer()

def preprocess(text):
    # Remove \n, \r, \t
    text = re.sub(r'[\n\r\t]+', ' ', text)

    # Remove HTML tags like <body>, but first remove script/style content
    soup = BeautifulSoup(text, 'html.parser')
    for tag in soup(['script', 'style', 'noscript']):
        tag.decompose()
    text = soup.get_text(separator=' ')
    
    # Lowercase
    text = text.lower()
    
    # Remove URLs
    text = re.sub(r'(?:https?://|ftp://|www\.)\S+|mailto:\S+|tel:\S+', '', text)
    
    # Remove emails
    text = re.sub(r'\b[a-z0-9._%+-]+@[a-z0-9.-]+\.[a-z]{2,}\b', '', text)
    
    # Remove phrases
    # for phrase in phrase_list:
    #     text = text.replace(phrase, ' ')

    # Remove non-ASCII characters
    text = re.sub(r'[^\x00-\x7F]', ' ', text)
    
    # Keep only letters and spaces
    text = re.sub(r'[^a-z\s]', '', text)

    # Tokenize and filter with stop words and lemmatization
    tokens = text.split()
    tokens = [lemmatizer.lemmatize(t) for t in tokens if len(t) >= 2 and t not in stop_words]
    
    return ' '.join(tokens)

In [3]:
# Load documents
documents = pd.read_csv('../data/english_pages_metadata_clean_with_labels.csv')
print(documents['manual_label_final'].value_counts())

# Apply preprocessing
documents['text_clean'] = documents['full_text'].apply(preprocess)


# Basic CountVectorizer
count_vec = CountVectorizer()
count_matrix = count_vec.fit_transform(documents['text_clean'])  # Fixed

print(f"Matrix shape: {count_matrix.shape}")
print(f"  - {count_matrix.shape[0]} documents")
print(f"  - {count_matrix.shape[1]} unique terms in vocabulary")

# Be careful with large vocabularies - this could print thousands of words
vocab = count_vec.get_feature_names_out()
print(f"\nFirst 20 terms: {vocab[:20]}")
print(f"Last 20 terms: {vocab[-20:]}")

doc_freq = np.asarray((count_matrix > 0).sum(axis=0)).flatten() # Number of documents that have each terms

print(f"Terms appearing in 1 doc: {(doc_freq == 1).sum()}")
print(f"Terms appearing in <5 docs: {(doc_freq < 5).sum()}")
print(f"Terms appearing in >50% docs: {(doc_freq > len(documents) * 0.5).sum()}")
print(f"Terms appearing in >85% docs: {(doc_freq > len(documents) * 0.85).sum()}")

manual_label_final
OTHER               158
BLOG                123
ECOMMERCE           114
EDUCATION            61
NEWS                 47
FORUM/DISCUSSION     46
TECHNICAL            34
GOVERNMENT           13
Name: count, dtype: int64


/tmp/ipykernel_74703/934885325.py:9: XMLParsedAsHTMLWarning: It looks like you're using an HTML parser to parse an XML document.

Assuming this really is an XML document, what you're doing might work, but you should know that using an XML parser will be more reliable. To parse this document as XML, make sure you have the Python package 'lxml' installed, and pass the keyword argument `features="xml"` into the BeautifulSoup constructor.

If you want or need to use an HTML parser on this document, you can make this warning go away by filtering it. To do that, run this code before calling the BeautifulSoup constructor:

    from bs4 import XMLParsedAsHTMLWarning
    import warnings

    warnings.filterwarnings("ignore", category=XMLParsedAsHTMLWarning)

  soup = BeautifulSoup(text, 'html.parser')


Matrix shape: (596, 43816)
  - 596 documents
  - 43816 unique terms in vocabulary

First 20 terms: ['aa' 'aaa' 'aaaa' 'aaaaa' 'aaaaaa' 'aaaaaaa'
 'aaacfbdburvncadkwicddxdqopvauvlooxgnevivzifwevajfhogfkaolrjuzbptdnxjmewmubhdorickodttxwq'
 'aaah' 'aaai' 'aac' 'aacaras' 'aachaara' 'aachaaraa' 'aacharas'
 'aadheenams' 'aadheenamsintroduction' 'aadhi' 'aaditya' 'aadiyaat' 'aaha']
Last 20 terms: ['zwart' 'zwarte' 'zweck' 'zwecke' 'zwecken' 'zweifel' 'zweifelsf' 'zweig'
 'zwischen' 'zwoelftekojelinkscom' 'zwolle' 'zxvcxv' 'zyban'
 'zygosaccharomyces' 'zygotorulaspora' 'zymoseptoria' 'zynga' 'zyxel' 'zz'
 'zzz']
Terms appearing in 1 doc: 28490
Terms appearing in <5 docs: 38151
Terms appearing in >50% docs: 4
Terms appearing in >85% docs: 0


After preprocessing, about 65% of the terms (28,490 out of 43,816) appeared in one document, which are most likely irrelavant words, which even includes typos. These terms were hard to be removed by the preprocessing which relies on regex.

Only four terms appeared in more than 50% of the documents, which are expected since the original data is not scraped based on a given topic. This also shows that the preprocessing has successfully removed sementically less meaningful words such as a, the, this, that, etc.

This information can be our guideline to set our max_df and min_df.

- TF-IDF Vectorizer

We start by setting max_df=0.5 and min_df=2 (documents fewer than 2) found from above. max_df=0.5 only removes 4 terms, but lowering max_df below that may remove some semantically meaningful terms, so we will stay at max_df=0.5

The TfidfVectorizer here is only used for the analysis purpose, but not for the actual model fitting.

In [4]:
tfidf_vec = TfidfVectorizer(
    max_df=0.5,      # Ignore terms in >85%
    min_df=2,         # Ignore terms in <5 docs
    # ngram_range=(1,2), # Unigrams and bigrams
    # max_features=10000  # Max vocabulary size
)
tfidf_matrix = tfidf_vec.fit_transform(documents['text_clean'])

print(f"TF-IDF Matrix shape: {tfidf_matrix.shape}")
print(f"Average non-zeros per document: {tfidf_matrix.nnz/tfidf_matrix.shape[0]}")

density = tfidf_matrix.nnz / (tfidf_matrix.shape[0] * tfidf_matrix.shape[1])
print(f"Sparsity: {(1-density) * 100:.4f}%")

TF-IDF Matrix shape: (596, 15322)
Average non-zeros per document: 228.6728187919463
Sparsity: 98.5076%


In [5]:
tfidf_vec = TfidfVectorizer(
    max_df=0.5,      # Ignore terms in >85%
    min_df=2,         # Ignore terms in <5 docs
    ngram_range=(1,2), # Unigrams and bigrams
    # max_features=10000  # Max vocabulary size
)
tfidf_matrix = tfidf_vec.fit_transform(documents['text_clean'])

print(f"TF-IDF Matrix shape: {tfidf_matrix.shape}")
print(f"Average non-zeros per document: {tfidf_matrix.nnz/tfidf_matrix.shape[0]}")

density = tfidf_matrix.nnz / (tfidf_matrix.shape[0] * tfidf_matrix.shape[1])
print(f"Sparsity: {(1-density) * 100:.4f}%")

TF-IDF Matrix shape: (596, 40472)
Average non-zeros per document: 345.2046979865772
Sparsity: 99.1471%


In [6]:
tfidf_vec = TfidfVectorizer(
    max_df=0.5,      # Ignore terms in >85%
    min_df=2,         # Ignore terms in <5 docs
    ngram_range=(1,2), # Unigrams and bigrams
    max_features=5000  # Max vocabulary size
)
tfidf_matrix = tfidf_vec.fit_transform(documents['text_clean'])

print(f"TF-IDF Matrix shape: {tfidf_matrix.shape}")
print(f"Average non-zeros per document: {tfidf_matrix.nnz/tfidf_matrix.shape[0]}")

density = tfidf_matrix.nnz / (tfidf_matrix.shape[0] * tfidf_matrix.shape[1])
print(f"Sparsity: {(1-density) * 100:.4f}%")


vocab = tfidf_vec.get_feature_names_out() # Term names
doc_freq = np.asarray((tfidf_matrix > 0).sum(axis=0)).flatten() # Number of documents that have each terms

n = 30
top_idx = doc_freq.argsort()[::-1][:n]

print("Most frequent terms (by document count):")
for i, idx in enumerate(top_idx, 1):
    print(f"  {i}. {vocab[idx]}: {doc_freq[idx]} docs")

TF-IDF Matrix shape: (596, 5000)
Average non-zeros per document: 183.85570469798657
Sparsity: 96.3229%
Most frequent terms (by document count):
  1. email: 263 docs
  2. right: 257 docs
  3. privacy: 252 docs
  4. policy: 245 docs
  5. content: 244 docs
  6. service: 239 docs
  7. time: 235 docs
  8. use: 219 docs
  9. copyright: 212 docs
  10. information: 209 docs
  11. term: 207 docs
  12. may: 207 docs
  13. one: 207 docs
  14. get: 204 docs
  15. privacy policy: 204 docs
  16. page: 197 docs
  17. post: 188 docs
  18. website: 188 docs
  19. work: 184 docs
  20. blog: 183 docs
  21. news: 182 docs
  22. like: 181 docs
  23. view: 181 docs
  24. first: 181 docs
  25. top: 180 docs
  26. year: 180 docs
  27. name: 179 docs
  28. best: 179 docs
  29. day: 175 docs
  30. help: 175 docs


## Part B: Classification

- Grid search for hyperparameter tuning and Cross validation.

In [9]:
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

scoring = {
    'accuracy': 'accuracy',
    'precision': make_scorer(precision_score, average='weighted', zero_division=0),
    'recall':    make_scorer(recall_score,    average='weighted', zero_division=0),
    'f1':        make_scorer(f1_score,        average='weighted', zero_division=0)
}

models = {
    'LogisticRegression': (
        LogisticRegression(max_iter=1000, class_weight='balanced'),
        {'clf__C': [0.1, 1.0, 5.0, 10.0]}
    ),
    'KNN': (
        KNeighborsClassifier(metric='cosine', n_jobs=-1),
        {'clf__n_neighbors': [3, 5, 7, 11], 'clf__weights': ['uniform', 'distance']}
    ),
    'SVM': (
        SVC(class_weight='balanced'),
        {'clf__C': [0.1, 1.0, 5.0], 'clf__kernel': ['linear', 'rbf']}
    ),
    'RandomForest': (
        RandomForestClassifier(class_weight='balanced', random_state=42, n_jobs=-1),
        {'clf__n_estimators': [100, 200], 'clf__max_depth': [None, 10, 20]}
    ),
    'DecisionTree': (
        DecisionTreeClassifier(class_weight='balanced', random_state=42),
        {'clf__max_depth': [5, 10, 15, None]}
    )
}

results = {}
for name, (model, params) in models.items():
    pipeline = Pipeline([
        ('tfidf', TfidfVectorizer(max_df=0.5, min_df=2, ngram_range=(1,2), max_features=5000)),
        ('clf', model)
    ])
    
    # Select the best model based on F1 score (refit='f1')
    gs = GridSearchCV(pipeline, params, cv=cv, scoring=scoring, refit='f1', n_jobs=-1)
    gs.fit(documents['text_clean'], documents['manual_label_final'])
    
    # Extract scores for the best parameter combination
    best_idx = gs.best_index_
    cv_results = gs.cv_results_
    
    results[name] = {
        'best_params': gs.best_params_,
        'accuracy':  cv_results['mean_test_accuracy'][best_idx],
        'precision': cv_results['mean_test_precision'][best_idx],
        'recall':    cv_results['mean_test_recall'][best_idx],
        'f1':        cv_results['mean_test_f1'][best_idx]
    }
    
    print(f"\n{name}")
    print(f"  Best params: {gs.best_params_}")
    print(f"  Accuracy:    {results[name]['accuracy']:.4f}")
    print(f"  Precision:   {results[name]['precision']:.4f}")
    print(f"  Recall:      {results[name]['recall']:.4f}")
    print(f"  F1:          {results[name]['f1']:.4f}")


LogisticRegression
  Best params: {'clf__C': 1.0}
  Accuracy:    0.5655
  Precision:   0.5661
  Recall:      0.5655
  F1:          0.5567

KNN
  Best params: {'clf__n_neighbors': 7, 'clf__weights': 'distance'}
  Accuracy:    0.4984
  Precision:   0.4962
  Recall:      0.4984
  F1:          0.4801

SVM
  Best params: {'clf__C': 1.0, 'clf__kernel': 'linear'}
  Accuracy:    0.5303
  Precision:   0.5999
  Recall:      0.5303
  F1:          0.5161

RandomForest
  Best params: {'clf__max_depth': 10, 'clf__n_estimators': 200}
  Accuracy:    0.5521
  Precision:   0.5660
  Recall:      0.5521
  F1:          0.5350

DecisionTree
  Best params: {'clf__max_depth': 15}
  Accuracy:    0.4397
  Precision:   0.4647
  Recall:      0.4397
  F1:          0.4330


- Report scores of each models.

In [10]:
summary = pd.DataFrame(results).T.drop(columns='best_params')
summary = summary.astype(float).round(4)
print("\n--- Summary ---")
print(summary.to_string())


--- Summary ---
                    accuracy  precision  recall      f1
LogisticRegression    0.5655     0.5661  0.5655  0.5567
KNN                   0.4984     0.4962  0.4984  0.4801
SVM                   0.5303     0.5999  0.5303  0.5161
RandomForest          0.5521     0.5660  0.5521  0.5350
DecisionTree          0.4397     0.4647  0.4397  0.4330
